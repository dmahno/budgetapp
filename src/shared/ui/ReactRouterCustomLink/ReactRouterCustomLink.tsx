import React from 'react';
import {LinkProps, useNavigate} from 'react-router-dom';
import LinkDS, {LinkProps as TLinkDSProperties} from 'antd/lib/typography/Link';

function isModifiedEvent(event: React.MouseEvent<HTMLAnchorElement>) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

interface IReactRouterCustomLinkProps extends Omit<LinkProps, 'to'> {
  to: string;
  disableAntColor?: boolean;
}

export const ReactRouterCustomLink = ({
  target,
  to,
  onClick,
  disableAntColor = false,
  ...props
}: IReactRouterCustomLinkProps & Partial<TLinkDSProperties>) => {
  const navigate = useNavigate();

  return (
    <LinkDS
      {...(props as TLinkDSProperties)}
      style={
        disableAntColor ? {color: 'inherit', textDecoration: 'none'} : undefined
      }
      onClick={(event: React.MouseEvent<HTMLAnchorElement>) => {
        try {
          if (onClick) onClick(event);
        } catch (ex) {
          event.preventDefault();
          throw ex;
        }

        if (
          !event.defaultPrevented &&
          event.button === 0 &&
          (!target || target === '_self') &&
          !isModifiedEvent(event)
        ) {
          event.preventDefault();
          navigate(to);
        }
      }}
    />
  );
};
