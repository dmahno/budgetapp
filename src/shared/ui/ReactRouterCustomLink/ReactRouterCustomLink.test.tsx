import {describe, it, expect, vi, beforeEach} from 'vitest';

import {render, screen, fireEvent} from 'shared/tests/test-utils';

import {ReactRouterCustomLink} from './ReactRouterCustomLink';

vi.mock('react-router-dom', async () => {
  const actual =
    await vi.importActual<typeof import('react-router-dom')>(
      'react-router-dom',
    );
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

const mockNavigate = vi.fn();
const mockOnClick = vi.fn();

describe('ReactRouterCustomLink', () => {
  beforeEach(() => {
    mockNavigate.mockClear();
    mockOnClick.mockClear();
  });

  it('renders correctly with default props', () => {
    render(<ReactRouterCustomLink to="/test">Test Link</ReactRouterCustomLink>);
    const linkElement = screen.getByText('Test Link');
    expect(linkElement).toBeInTheDocument();
    expect(linkElement).toHaveStyle({color: 'rgb(9, 88, 217)'});
  });

  //TODO нужно сделать disable
  //   it('applies style when disableAntColor is true', () => {
  //     render(
  //       <ReactRouterCustomLink to="/test" disableAntColor>
  //         Styled Link
  //       </ReactRouterCustomLink>,
  //     );
  //     const linkElement = screen.getByText('Styled Link');
  //     expect(linkElement).toHaveStyle({color: 'inherit', textDecoration: 'none'});
  //   });

  it('calls navigate function on click', () => {
    render(
      <ReactRouterCustomLink to="/navigate-test">
        Navigate Link
      </ReactRouterCustomLink>,
    );
    const linkElement = screen.getByText('Navigate Link');
    fireEvent.click(linkElement);
    expect(mockNavigate).toHaveBeenCalledWith('/navigate-test');
  });

  it('prevents navigation on modified event', () => {
    render(
      <ReactRouterCustomLink to="/navigate-test">
        Modified Event Link
      </ReactRouterCustomLink>,
    );
    const linkElement = screen.getByText('Modified Event Link');
    fireEvent.click(linkElement, {ctrlKey: true});
    expect(mockNavigate).not.toHaveBeenCalled();
  });

  it('calls onClick and then navigates if not prevented', () => {
    render(
      <ReactRouterCustomLink to="/click-test" onClick={mockOnClick}>
        Clickable Link
      </ReactRouterCustomLink>,
    );
    const linkElement = screen.getByText('Clickable Link');
    fireEvent.click(linkElement);
    expect(mockOnClick).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith('/click-test');
  });
  it('matches the snapshot', () => {
    const {asFragment} = render(<ReactRouterCustomLink to="/click-test" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
