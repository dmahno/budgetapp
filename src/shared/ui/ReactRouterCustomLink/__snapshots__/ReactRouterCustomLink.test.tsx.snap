// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReactRouterCustomLink > matches the snapshot 1`] = `
<DocumentFragment>
  <a
    class="ant-typography css-dev-only-do-not-override-ni1kz0"
  />
</DocumentFragment>
`;
